store.go

```diff
@@ -1,174 +1,214 @@
+		// until the object is Closed/Sync'ed.
+		NewObject(o string) (Object, error)
+
+		// Delete removes the object from the cloud store.
+		Delete(o string) error
 	}
 
-	switch csctx.TokenSource {
-	case GCEDefaultOAuthToken, GCEMetaKeySource, LyticsJWTKeySource, GoogleJWTKeySource:
-		googleclient, err := NewGoogleClient(csctx)
-		if err != nil {
-			return nil, err
-		}
-		return gcsCommonClient(googleclient.Client(), csctx)
-	case LocalFileSource:
-		prefix := fmt.Sprintf("%s:", csctx.LogggingContext)
-		l := LogConstructor(prefix)
-
-		store, err := NewLocalStore(csctx.LocalFS, csctx.TmpDir, l)
-		if err != nil {
-			l.Errorf("error creating the store. err=%v ", err)
-			return nil, err
-		}
-		return store, nil
-	default:
-		return nil, fmt.Errorf("bad sourcetype: %v", csctx.TokenSource)
+	// Object is a handle to a cloud stored file/object.  Calling Open will pull the remote file onto
+	// your local filesystem for reading/writing.  Calling Sync/Close will push the local copy
+	// backup to the cloud store.
+	Object interface {
+		Name() string
+		String() string
+		Updated() time.Time
+		MetaData() map[string]string
+		SetMetaData(meta map[string]string)
+		// StorageSource is the type of store.
+		StorageSource() string
+		// Open copies the remote file to a local cache and opens the cached version
+		// for read/writing.  Calling Close/Sync will push the copy back to the
+		// backing store.
+		Open(readonly AccessLevel) (*os.File, error)
+		// Release will remove the locally cached copy of the file.  You most call Close
+		// before releasing.  Release will call os.Remove(local_copy_file) so opened
+		// filehandles need to be closed.
+		Release() error
+		// Implement io.ReadWriteCloser Open most be called before using these
+		// functions.
+		Read(p []byte) (n int, err error)
+		Write(p []byte) (n int, err error)
+		Sync() error
+		Close() error
+
+		// Delete removes the object from the cloud store.
+		Delete() error
 	}
-}
 
-func gcsCommonClient(client *http.Client, csctx *CloudStoreContext) (Store, error) {
-	project := csctx.Project
-	bucket := csctx.Bucket
-	prefix := fmt.Sprintf("%s:(project=%s bucket=%s)", csctx.LogggingContext, project, bucket)
-	l := LogConstructor(prefix)
+	// ObjectIterator interface to page through objects
+	ObjectIterator interface {
+		Next() (Object, error)
+	}
 
-	gcs, err := storage.NewClient(context.Background(), option.WithHTTPClient(client))
-	if err != nil {
-		l.Errorf("%v error creating Google cloud storage client. project:%s gs://%s/ err:%v ",
-			csctx.LogggingContext, project, bucket, err)
-		return nil, err
+	// Objects are just a collection of Object(s).
+	// Used as the results for store.List commands.
+	Objects []Object
+
+	// TokenSource Is the source/location/type of token
+	TokenSource string
+
+	// AccessLevel is the level of permissions on files
+	AccessLevel int
+
+	// Config the cloud store config parameters
+	Config struct {
```

Comment:
What if we made `Config` and interface with a single func `Type()`.   Then each store type can have a concrete implementation for `Config`.   Then the store's constructor can cast the interface to it's concrete type, to extract fields like `JwtConf`, `Project`,  `TokenSource`, etc...

Reply:
I don't mind checking our your PR and making that change if agree.

Reply:
Or do we do something more like this?  https://github.com/araddon/qlbridge/blob/master/datasource/files/filesource.go#L136

https://github.com/araddon/qlbridge/blob/master/schema/schema.go#L166

Or are we thinking about this right?   Do we need Configuration/Factory or do we go with package specific implementation without the central/global factory?
```go
google.NewStore(google-specific-params)
```

Reply:
and yes, go for sending/pushing straight to this pr if you want, or different pr into it.